components/sync15/src/bso/crypto.rs

```diff
@@ -0,0 +1,195 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Support for "encrypted bso"s, as received by the storage servers.
+//! This module decrypts them into IncomingBso's suitable for use by the
+//! engines.
+use super::{IncomingBso, IncomingEnvelope, OutgoingBso, OutgoingEnvelope};
+use crate::error;
+use crate::key_bundle::KeyBundle;
+use crate::EncryptedPayload;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+// The BSO implementation we use for encrypted payloads.
+// Note that this is almost identical to the IncomingBso implementations, except
+// instead of a String payload we use an EncryptedPayload. Obviously we *could*
+// just use a String payload and transform it into an EncryptedPayload - any maybe we
+// should - but this is marginally optimal in terms of deserialization.
+#[derive(Deserialize, Debug)]
+pub struct IncomingEncryptedBso {
+    #[serde(flatten)]
+    pub envelope: IncomingEnvelope,
+    // pub is slightly unfortuate, but we need it so we can get hold of our global keys.
+    #[serde(
+        with = "as_json",
+        bound(deserialize = "EncryptedPayload: DeserializeOwned")
+    )]
+    payload: EncryptedPayload,
+}
+
+impl IncomingEncryptedBso {
+    pub fn new(envelope: IncomingEnvelope, payload: EncryptedPayload) -> Self {
+        Self { envelope, payload }
+    }
+    /// Decrypt a BSO, consuming it into a clear-text version.
+    pub fn into_decrypted(self, key: &KeyBundle) -> error::Result<IncomingBso> {
+        Ok(IncomingBso::new(self.envelope, self.payload.decrypt(key)?))
+    }
+    /// Reference the payload.
+    pub fn payload(&self) -> &EncryptedPayload {
+        &self.payload
+    }
+    /// Consume the payload
+    pub fn consume_payload(self) -> EncryptedPayload {
```

Comment:
We don't necessarily need to follow this but https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv (also a generally great read)

would imply that `into_payload()` makes sense since it consumes the data. However `consume_payload` is also pretty clear hah.

Reply:
As above, I killed these methods and just made the payload `pub(crate)`